types/events/watcher.go

```diff
@@ -0,0 +1,132 @@
+package events
+
+import (
+	goevents "github.com/docker/go-events"
+	"golang.org/x/net/context"
+)
+
+// MatcherFunc return true if the event matches or false if not.
+type MatcherFunc func(event Message) bool
+
+// Watcher watches an event stream from the daemon providing
+// an easy way to filter events to specific channels
+type Watcher interface {
+	// Watch all the events from the daemon.
+	//
+	// If matcher functions are specified then the channel will only
+	// receive events that match any of the matchers.
+	// For example, say you pass in two matchers: one for
+	// label: foo and the other for label: bar. The channel returned
+	// will receive events that have either label.
+	//
+	// Cancel the context to quit watching the events on this channel
+	Watch(ctx context.Context, matcher ...MatcherFunc) <-chan Message
+}
+
+type watcher struct {
+	broadcast *goevents.Broadcaster
+	events    <-chan Message
+	buffer    int
+	context   context.Context
```

Comment:
Let's avoid putting `context` on the struct. Let's use a `closed` channel to signify shutdown.